### DISCUSSION TITLE
Restrict SELECT operation to users within the same organisation - issue with auth.uid() function

### DISCUSSION THREAD

**Discussion by ChaddRobertson** [supabase/supabase#15927] [Questions] [answered]
Hi everyone, I've been stuck on implementing this RLS policy for a while now.

I've created an application where users must register using an organisation key (a uuid that is created when a new organisation is added to the `organisations` table). The registration of a new user triggers a member instance to be created in the `members` table, where the `organisation_id` is automatically populated from the new `auth.users` instance's metadata and references the `organisation` table.

My table definitions are as follows:

The organisations table:

```
create table
  public.organisations (
    id uuid not null default uuid_generate_v4 (),
    created_at timestamp without time zone not null default now(),
    company_email character varying(150) not null,
    company_name character varying(150) not null,
    administrator_email character varying(150) not null,
    administrator_first_name character varying(150) not null,
    administrator_last_name character varying(150) not null,
    total_seats integer not null,
    available_seats integer not null,
    constraint organisations_duplicate_pkey primary key (id)
  ) tablespace pg_default;
```

The members table:

```
create table
  public.members (
    id uuid not null,
    organisation_id uuid not null,
    email character varying not null,
    last_name character varying not null,
    first_name character varying not null,
    job_title character varying null,
    employee_number character varying null,
    dark_theme boolean not null default false,
    user_color character varying(11) not null default generate_random_pastel_color (),
    constraint members_pkey primary key (id),
    constraint members_id_fkey foreign key (id) references auth.users (id) on delete cascade,
    constraint members_organisation_id_fkey foreign key (organisation_id) references organisations (id) on delete cascade
  ) tablespace pg_default;
```

Its important that members may only be able to view information of other members if they share the same organisation_id, so I would like to create a row level security policy that restricts access like this, but I am hitting a road block - whenever I attempt to read my data client side, I get NULL returned:

![Capture](https://github.com/supabase/supabase/assets/69690158/092834f9-e573-4fc2-a79c-dcfb1b51fd8d)

I know that the issue is with my policy, because everything works as intended when I ommit it and filter the organisation members client side, like so:

```
static Future<List<MemberModel>> getMemberModelsOfUserOrganisation({required SupabaseClient supabaseClient, required MembersSupabaseTable membersSupabaseTable}) async {

    MemberModel currentMemberModel = await getUserMemberModel(supabaseClient: supabaseClient, membersSupabaseTable: membersSupabaseTable);
    String currentMemberOrganisationID = currentMemberModel.organisationID;

    final List<dynamic> response = await supabaseClient
        .from(membersSupabaseTable.tableName)
        .select()
        .eq(membersSupabaseTable.organisationID, currentMemberOrganisationID);

    return response.map((map) => MemberModel.fromMap(map)).toList();
  }
```

My (non-functional) policy is defined for public roles as follows, and is based on the [postgres docs](https://www.postgresql.org/docs/current/ddl-rowsecurity.html):

```
CREATE POLICY member_select_policy ON public.members
    FOR SELECT
    USING (organisation_id = (SELECT organisation_id FROM public.members WHERE id = auth.uid()));
```

Strangely, the boolean check works fine if I manually specify the uuid, rather than using `auth.uid()`, where the following indeed returns `true`:

```
SELECT '24cf5dec-a570-4ffd-8188-be01f9cfdf18' = (SELECT organisation_id FROM public.members WHERE members.id = '3789aba9-c72f-4f3d-ba11-597b052e30f2');
```

How would I define a RLS policy that only allows members to SELECT other members if and only if they share an organisation_id with the current user? Adjusting the applicable roles to that of "authenticated" does not yield a change.

I'm pulling my hair out trying to find the issue with this.

**Comment by GaryAustin1** [comment, marked as answer]
If I'm reading your code correctly you are doing a select policy on the same table you use in the select policy.  That won't work as to access the table you already need select permission.   

You will need to do your test in a security definer function so the actual test bypasses RLS.
An example: https://supabase.com/docs/guides/auth/row-level-security#policies-with-security-definer-functions

**Reply by ChaddRobertson** [reply]
This makes a mountain of sense. I had a thought that my mistake may have been tied to something along these lines, but didn't quite have the experience to have anything more than a gut feeling. Will give this a crack and close out if successful. Cheers for the suggestion.
